### PR TITLE
Infer parameter names and types when applying Delcare Method codefix 

### LINKED ISSUE #22180: Declare method refactor/quickfix should fill in proper parameter types
The "declare method" quick fix should not declare all parameter (and return type) as `any`, but should instead enter the proper argument types:

```typescript
class A { 
}
var a = new A();
a.foo(123, "qwerty");
  ~~~ 
```
Declare method 'foo' should yield:

```typescript
class A {
    foo(arg0: number, arg1: string) {
        throw new Error("Method not implemented");
    }
}
``` 

Rather than the current behaviour of using `any` in all positions.

### PR DESCRIPTION
**Pull request by markusjohnsson**
Uses argument types from call site to define parameter types, after having widened literal types, in the same way as is done in the Extract Symbol refactor.

Uses argument identifier or property names from call site to define parameter names, or `argN` (as before) if no names are available at the call site:

```typescript
const x = 1;
const y = "test";
m.m0(x, y, false);

    // result of  Declare Method 'm0':
    m0(x: number, y: string, arg2: boolean) : any {
        throw new Error("Method not implemented.");
    }
```

Notes:
 * IMO, it would be better to not explicitly define a return type, than to define `any`
 * Empty array literals yields `undefined[]` as type, which, while not wrong, might not be ideal either.

Fixes #22180 

Edit: Added error throwing in the example, which has not changed in this PR.

### CODE REVIEW

**Comment by markusjohnsson** [comment]
If my changes caused `tests/cases/compiler/largeControlFlowGraph.ts` to timeout, I would need some help to look at it. But I doubt that it was my changes that caused it. Is it possible to just re-run the test?

File: src/services/codefixes/helpers.ts [hunk 1]
```diff
@@ -107,7 +107,18 @@ namespace ts.codefix {
         return nodes && createNodeArray(nodes.map(getSynthesizedDeepClone));
     }
 
-    export function createMethodFromCallExpression({ typeArguments, arguments: args }: CallExpression, methodName: string, inJs: boolean, makeStatic: boolean): MethodDeclaration {
+    export function createMethodFromCallExpression(context: CodeFixContextBase, { typeArguments, arguments: args }: CallExpression, methodName: string, inJs: boolean, makeStatic: boolean): MethodDeclaration {
+        const checker = context.program.getTypeChecker();
+        const types = map(args,
+            (arg) => {
```
Comment (DanielRosenwasser) [inline src/services/codefixes/helpers.ts]: I'd delegate this to a linter, but looks like it's not enabled here, so but no parens around single parameters in arrow functions

**Comment by markusjohnsson** [comment]
Thank you, @DanielRosenwasser . I've removed the parens. Somehow a test timed out again. When adding commits, should I squash all changes into one or is it OK lite this?

**Comment by markusjohnsson** [comment]
@DanielRosenwasser : Do you need me to do anything more with this before it can be merged?

**Comment by mhegazy** [comment]
thanks!

**Comment by ahocevar** [comment]
Trying to understand this pull request: where has the functionality of `src/compiler/declarationEmitter.ts` been moved to when it was deleted in cb5b773? The reason why I'm asking is because I'm trying to take another stab at #7546.

**Pull request merged**